### ISSUE TITLE
NS lookup during TXT record validation - DnsPreCheck issue

### ISSUE THREAD

**Issue by EugenMayer** [xenolf/lego#544]
AFAIU lego tries, in difference to certbo / acme.sh, to avoid the issue with the static "timeout" setting on DNS-01 based challenged.

Instead of waiting for X seconds and the notify LE to try to find the TXT record, i understand that you guys try to check for the TXT record yourself?

While you do that, you try to find the NS server for the domain of e.g. foo.domain.tld.. and then query the primary namesever for the TXT record and wait until you find it - and then notify LE to check themselfs

That all sounds well designed, until we hit the actual sweet spot of DNS-01 - isolated networks ( right? )
Since that is the main reason to actually use DNS-01 when your public domains are used in a LAN and there are no public DNS entries for your subdomains at all .. in addition that you do not want to open your firewall for http challenge in addition ( but that a soft spot ).

In those isolated / LAN networks, you will have a local DNS server. And at least for us with OPNsense + unbound, it will not return a NS entry for a private domain (public tld). Thus, any usage of LEGO fails, since not primary server can be found and the DNS precheck process fails. 

I have seen `--dns-resolvers value` and it defaults to google - i guess to mitigate this issue right? Why i am asking is, that the load-balancer traefik does implement `lego` but i am not sure how they handle this parameter and if it is related at all.

Can you guys explain your mind with this particular issue and how you think this should be solved in general and as libarary?

Thanks

**Comment by EugenMayer** [comment]
Any ways of getting a response here? using `certbot` or `acme.sh` all this works without any issues, so its really just `LEGO` related. Any way we can create progress here?

**Comment by captncraig** [comment]
In order to solve the DNS-01 challenge, you need publicly accessible nameservers, right? You keep talking about SOA records, and that is not quite accurate. What lego looks for are the `NS` records for your domain. The ones that get registered with the tld nameservers. It then queries those to make sure the appropriate TXT records are present.

What nameservers are you using for these "private domains"? Are they publicly accessible? I'm just trying to understand what circumstances lead to nameservers being accessible by Let's Encrypt but not the Lego client.

**Comment by EugenMayer** [comment]
@captncraig sure, let me rephrase - let me know if i still miss something / it sounds confusion. I changed my SOA assumption to NS - but it does not matter, it is the same thing in this regard

Let us assume, we have a public domain called 'mycompany.com':

 a) its primary/secondary DNS servers are lars.clouflare.com 
 b) the subdomain `internal.mycompany.com` is used to resolve private IPs only - its used internally
 c) there is no way to provide http based challenge due to b)
 d) there is no DNS entry for `internal.mycompany.com` other then an A record, SOA for `mycompany.com`

Lets assume, we have a company private network with the network 10.1.7.0/24:
 i) it has a DNS server on 10.1.7.1 which is also the GW
 ii) this DNS server can recurse for public lookup to 8.8.8.8
 iii) All clients in the company network are using 10.1.7.1 as their only DNS server
 iv) This DNS server is a unbound server hosted on https://opnsense.org
 v) this DNS server defines `internal.mycompany.com` as a private domain .. https://www.unbound.net/documentation/unbound.conf.html ... search for `private-domain: <domain name>`

When resolving `dig @10.1.7.1 internal.mycompany.com NS` i get this response

```
dig @10.1.7.1 neptung.kw.kontextwork.com NS                  [18:49:13]

; <<>> DiG 9.8.3-P1 <<>> @10.1.7.1 internal.mycompany.com NS
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 12184
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;internal.mycompany.com.	IN	NS

;; Query time: 839 msec
;; SERVER: 10.1.7.1#53(10.1.7.1)
;; WHEN: Tue May 29 18:49:56 2018
;; MSG SIZE  rcvd: 44
```

when i do SOA lookup for `mycompany.com` using googles DNS i get

`dig @8.8.8.8 mycompany.com`

```
; <<>> DiG 9.8.3-P1 <<>> mycompany.com NS
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 14177
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;mycompany.com.		IN	NS

;; ANSWER SECTION:
mycompany.com.	21599	IN	NS	lars.ns.cloudflare.com.
mycompany.com.	21599	IN	NS	yolanda.ns.cloudflare.com.

;; Query time: 50 msec
;; SERVER: 10.10.10.1#53(10.10.10.1)
;; WHEN: Tue May 29 18:49:13 2018
;; MSG SIZE  rcvd: 88
```

**Problem**:

When Traefik - LEGO - tries to DNS-01 challenge `something.internal.mycompany.com` it will ask the primary DNS server of the Traefik server, which is `10.1.7.1` for a NS and fails ( see above ) and this fails the challenge.

But its even worse, due to #551 i even fails when i manually change the namservers to

```
namservers 8.8.8.8
namservers 10.1.7.1
```

since again, somehow the second DNS server gets used, see #551 for more details.

---

Did that answer your questions?

Hint: all answers are real, i just replaced the real domain with mycompany.com for privacy. We can have  a chat somewhere if you want the real domains in do some DNS yourself.

Hint2: with `certbot`/`acme.sh` both work without any issues on the very same traefik server with 10.1.7.1 as DNS, with CF as provider and such.

**Comment by EugenMayer** [comment]
@captncraig any way you could answer on this / we could create some progress here? That is a real blocker ;/ Thanks!

**Comment by EugenMayer** [comment]
solved it on my side, anybody running into that, look for `typetransparent` to get arround this mess
